### ISSUE TITLE
I cannot understand how to use the client. Created a main.cpp but it won't work

### ISSUE THREAD

**Issue by NinaGitelman** [OpenCoreCH/TCPunch#1]
This is what I used for a main on the peers:
* Ive managed to connect to the server from the clients but not between 

peer1:
#include "tcpunch.h"
#include <iostream>

using std::cout;
using std::endl;

int main()
{
    char buffer[1024];
    int sock_fd = pair("my_pairing_name", "my-ip");
    std::cout<<"here!";
    int n = recv(sock_fd, buffer, 1024, 3126);
    std::cout << "buffer: "<< buffer << endl; 

    

}

-------------------------------------------------------------
// peer2:
#include "tcpunch.h"
#include <iostream>

using std::cout;
using std::endl;

int main()
{
    char buffer[1024];
    sock_fd = pair("my_pairing_name", "myip");
    char sendBuffer[] = "Hello world!";
    int bytes = send(sock_fd, buffer, 14, 0);

}

**Comment by mcopik** [comment]
@Ninavwx Hi! Thanks for your interest in the project!

Can you please explain a little bit about where you are running the code and what the problem is exactly? Do you see an error?

**Comment by NinaGitelman** [comment]
I ran the server code in an ec2 machine (server instance from amazon) and
the client code in two different computers running ubuntu linux in
different networks, connected to the internet.

I managed to connect the clients to the server but not to each other

It does not throw any error, just simply doesn't send a message from one
computer to the other. I also scanned the network traffic and checked for
upcoming packages from that ip. There were none.

I am not sure what could be the problem or even if my code for the client
sending and receiveing a message (sent in the issue) is correct.

I would be glad to receive help for this issue.

Thank you for your time!

**Comment by mcopik** [comment]
@Ninavwx Whether TCP hole punching will work depends on the NAT implementation of networks. It can be possible that between two networks that you are using, no hole punching is possible (details are in the thesis linked in README); or that our implementation is somehow insufficient.

Based on your description it is likely that no messages are sent, and clients are stuck in `pair` method as they cannot establish a connection.

Perhaps @OpenCoreCH has some suggestions? :-)

**Comment by OpenCoreCH** [comment]
Yes the code generally looks correct, but it is possible that the networks do not support TCP hole punching because of the port allocation scheme of the routers or their behaviour to unexpected SYN packets (see https://en.wikipedia.org/wiki/TCP_hole_punching for details). If it is caused by the latter, UDP hole punching may still be succesful, so trying out something like https://github.com/ckennelly/hole-punch could be worthwhile.

**Comment by schaumb** [comment]
I noticed that when SO_REUSEADDR is set, the connection binds to the public port:
https://github.com/OpenCoreCH/TCPunch/blob/main/client/tcpunch.cpp#L151

However, I believe that instead of using the public port (public_info.port), it would be more appropriate to use the already-bound local port. This is because the NAT may assign a different public port than the one used locally.

Suggested change:
```
    sockaddr_in local_address{};
    socklen_t address_len = sizeof(local_address);

    if (getsockname(socket_rendezvous, (struct sockaddr*)&local_address, &address_len) == -1) {
        error_exit("getsockname failed");
    }

    local_port_addr.sin_port = local_address.sin_port; // replacement of public_info.port;
```

Or am I misunderstanding the intended behavior here?


----------

The same port is also passed to the peer_listen thread, as seen here:
https://github.com/OpenCoreCH/TCPunch/blob/main/client/tcpunch.cpp#L117
https://github.com/OpenCoreCH/TCPunch/blob/main/client/tcpunch.cpp#L40

In this context, I believe it would be sufficient to pass only the local port to the thread, rather than the entire PeerConnectionData structure.

**Comment by mcopik** [comment]
@schaumb Thanks for the suggestion! Did you notice the issue when reading the code, or did you find this problem while debugging?

**Comment by schaumb** [comment]
I came across the issue while reading through the code. I was looking for a straightforward, dependency-free C++ implementation of TCP hole punching after I studied the method, and I found this project.

**Comment by mcopik** [comment]
@schaumb Thank you! I will have to investigate this and see if this changes behavior.
